java/arcs/core/analysis/DependencyNode.kt

```diff
@@ -23,117 +23,210 @@ private typealias Path = List<Identifier>
- *     output: writes PetStats {
- *       ratio: inline Domestication {trained: Number, total: Number},
- *       family: Text,
- *       limit: Number
- *     } = new PetStats {
- *       ratio: new Domestication {
- *         trained: input.cats,
- *         total: input.cats + input.dogs,
- *       },
- *       family: input.household,
- *       limit: 5
- *     }
- *   ```
+ *  ```
+ *  particle Example
+ *    input: reads Foo {x: Number, y: Number, z: Number}
+ *    output: writes Bar {a, b} =
+ *      from f in input
+ *      where f.x > 10
+ *      select new Bar {
+ *        a: f.x
+ *        b: f.y + f.z
+ *      }
+ *  ```
  *
- *   This can be translated to:
+ *  This translates to:
  *
- *   ```
- *   DependencyNode.AssociationNode(
- *    "ratio" to DependencyNode.AssociationNode(
- *      "trained" to DependencyNode.Input("input", "cats"),
- *      "total" to DependencyNode.DerivedFrom(
- *        DependencyNode.Input("input", "cats"),
- *        DependencyNode.Input("input", "dogs")
- *      )
+ *  ```
+ *  DependencyNode.Nodes(
+ *    DependencyNode.Equal(
+ *      "a",
+ *      dependency = setOf(DependencyNode.Equal("input", "x")),
+ *      influence = setOf(DependencyNode.Derived("input", "x"))
  *    ),
- *    "family" to DependencyNode.Input("input", "household"),
- *    "limit" to DependencyNode.LITERAL
- *   )
- *   ```
- *
- *   We can represent this DAG graphically:
+ *    DependencyNode.Equal(
+ *      "b",
+ *      dependency = setOf(
+ *        DependencyNode.Derived("input", "y"),
+ *        DependencyNode.Derived("input", "z")
+ *      ),
+ *      influence = setOf(DependencyNode.Derived("input", "x"))
+ *    )
+ *  )
+ *  ```
  *
- *                    (root)
- *            __________|______
- *           /      |          \
- *          /      /        _(ratio)_
- *        /       /        |         \
- *   (limit)  (family)  (trained)  (total)
- *     |        |          |     /   |
- *     |        |          |   /     |
- *     x     (household)  (cats)  (dogs)
+ *  Graphically, this is represented as:
  *
- *   This internal representation, in turn, can be translated into the following claims:
+ *  >  (input) --> ((x))==========
+ *  >     \                   \\ \\
+ *  >      \                  ||  V
+ *  >       \--> (x) <--------V--(a)-|[Bar]
+ *  >        \---> ((y)) <--(b)------|
+ *  >         \---> ((z)) <-/
  *
- *   ```
- *   claim output.ratio.trained derives from input.cats
- *   claim output.ratio.total derives from input.cats and derives from input.dogs
- *   claim output.family derives from input.household
- *   ```
+ *  This internal representation, in turn, can be translated into the following claims:
+ *  ```
+ *  claim output.a is equal to input.x and is influenced by input.x
+ *  claim output.b is derived from input.y and is derived from input.z and is influenced by input.x
+ *  ```
  */
 sealed class DependencyNode {
```

Comment:
A high-level comment. I was hoping that we could structure our classes as follows:
 - The base `DependencyNode` with `path`, `dependency`, and `influencedBy` as members.
 - `AssociationNode` as before
 - `Derived` node with a single `DependencyNode` or a `Set<DependencyNode>`
 - `Input` would be represented as an `AssociationNode` with empty `associations`.

Reply:
Something I don't quite yet grok: In this proposal, why have a `dependency` field at all? If `AssociationNode` is the same as before, then dependencies are implied by pairs / entries in the map. Under the current PR, pairs are converted into nodes with `dependency` edges (they actually form a graph).